src/llama.cpp

```diff
@@ -20462,6 +20462,10 @@ void llama_kv_cache_update(struct llama_context * ctx) {
     llama_kv_cache_update_internal(*ctx);
 }
 
+bool llama_kv_cache_can_shift(struct llama_context * ctx) {
+    return ctx->model.arch != LLM_ARCH_DEEPSEEK2; // not supported due to MLA
+}
```

Comment:
Should this return false for recurrent models as well? Not sure what's the logic there, but `llama_kv_cache_update_internal` silently ignores models with `LLAMA_ROPE_TYPE_NONE`.

Reply:
Yes, it's likely needed to return false for recurrent models.

The reason to do nothing in `llama_kv_cache_update_internal` when the rope type is none is because when we apply shifts to the KV cache using functions like `llama_kv_cache_seq_add()`, we do 2 things:

- Update the positions of the KV cells - i.e. just modify the meta data in `llama_kv_cell`
- Re-rope the data in the KV cells

The later step is necessary only if the data is roped. For ALiBi models for example, we should not apply this second step, but in theory we still support "shifting" the KV cache for those models, since the positional information is in the KQ mask.